src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+            }
+          } else {
+            // Validation is still in progress, so wait for a while and try again.
+            validationCheckTimeout = setTimeout(
+              pollValidationStatus,
+              this.validationCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollValidationStatus();
+    });
+  }
+
+  async doNewAddonSubmit(uuid: string, metaDataJson: Object): Promise<any> {
+    const url = `${this.apiUrlPrefix}addon/`;
+    const jsonData = { version: { upload: uuid }, ...metaDataJson };
+    const response = await this.fetch(url, 'POST', JSON.stringify(jsonData));
+    return this.getJson(response);
+  }
+
+  doNewAddonOrVersionSubmit(
+    addonId: string,
+    uuid: string,
+    metaDataJson: Object,
+  ): Promise<typeof Response> {
+    const url = `${this.apiUrlPrefix}addon/${addonId}/`;
+    const jsonData = { version: { upload: uuid }, ...metaDataJson };
+    return this.fetch(url, 'PUT', JSON.stringify(jsonData));
+  }
+
+  waitForApproval(addonId: string, versionId: number): Promise<string> {
+    let approvalCheckTimeout;
+    const detailUrl =
+      `${this.apiUrlPrefix}addon/${addonId}/versions/${versionId}/`;
+
+    return new Promise((resolve, reject) => {
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(approvalCheckTimeout);
+
+        reject(new Error('Approval Timeout.'));
+      }, this.approvalCheckTimeout);
+
+      const pollApprovalStatus = async () => {
+        try {
+          const detailResponse = await this.fetch(detailUrl);
+          if (!detailResponse.ok) {
+            return reject(new Error('Getting addon details failed.'));
+          }
+          const detailResponseData = await detailResponse.json();
+
+          const file = detailResponseData.file;
+          if (file && file.status === 'public') {
+            clearTimeout(abortTimeout);
+            resolve(file.url);
+          } else {
+            // The add-on hasn't been approved yet, so wait for a while and try again.
+            approvalCheckTimeout = setTimeout(
+              pollApprovalStatus,
+              this.approvalCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollApprovalStatus();
+    });
+  }
+
+  async getJson(response: typeof Response): Promise<any> {
+    if (response.status < 100 || response.status >= 500) {
+      return new Promise((resolve, reject) => {
+        reject(new Error(`Getting response failed: ${response.status}.`));
+      });
+    } else {
+      const data = await response.json();
+      return new Promise((resolve, reject) => {
+        if (!response.ok) {
+          this.logger.log(data);
+          reject(new Error('Bad Request.'));
+        } else {
+          resolve(data);
+        }
+      });
+    }
```

Comment:
There was a reason I created two promises... but I can't remember now what problem it solved!

Reply:
> * which case would make await response.json() to don't throw and response.ok to be false, I guess it may be because we got a client error status along with a more detailed description of the issue in JSON format, is that the case?

`response.ok` would be false if it's a 400 error, which we use to return property validation, and that's all in JSON format.

> * should we also take into account a await response.json() call that throw (which should only happen when the response isn't in a valid JSON format, may be actually even more unexpected but maybe worth to be checked explicitly and maybe even log the result in text form

`response.json` could throw, but it'd be something exceptional - a server error, etc - that lead to a non-JSON response. (addons-server _should_ always return a json response for an api request but server errors are unpredictable)

> If that is the case, then we should log that as a debug log and use the log.debug (and the logger created using the createLogger helper as described in another review comment).

I've not been very consistent at logging, I know.  I've pretty much assumed that if the developer strays off the happy path of 200/201 and 400 validation errors then the tool will just error and the Promise rejections will show enough in the console to be able to log an issue with us.  But if there is something/somewhere in particular I should add logging then I can do that.